js/common/bootstrap/modules/buffer.js

```diff
@@ -162,109 +170,99 @@ global.DEFINE_MODULE('buffer', (function () {
-          // We are a string
-          this.length = this.write(subject, 0, encoding);
-        // if subject is buffer then use built-in copy method
-        } else if (Buffer.isBuffer(subject)) {
-          if (subject.parent)
-            subject.parent.copy(this.parent,
-              this.offset,
-              subject.offset,
-              this.length + subject.offset);
-          else
-            subject.copy(this.parent, this.offset, 0, this.length);
-        } else if (isArrayIsh(subject)) {
-          for (var i = 0; i < this.length; i++)
-            this.parent[i + this.offset] = subject[i];
-        }
-      }
+  function createBufferFromString(subject, encoding) {
+    const buffer = Object.create(Buffer.prototype);
+    const length = Buffer.byteLength(subject, encoding);
+    buffer.parent = new SlowBuffer(length);
+    buffer.offset = 0;
+    buffer.length = buffer.parent.write(subject, encoding);
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
+
+  function createBufferFromArrayBuffer(subject, byteOffset = 0, length = subject.byteLength - byteOffset) {
+    const buffer = Object.create(Buffer.prototype);
+    const dv = new DataView(subject);
+    buffer.parent = new SlowBuffer(length);
+    buffer.offset = 0;
+    buffer.length = length;
+    for (let i = 0; i < length; i++) {
+      buffer.parent[i] = dv.getUint8(byteOffset + i);
     }
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
 
+  function createBufferFromByteArray(subject) {
+    const buffer = Object.create(Buffer.prototype);
+    buffer.parent = new SlowBuffer(subject.length);
+    buffer.offset = 0;
+    buffer.length = subject.length;
+    for (let i = 0; i < subject.length; i++) {
+      buffer.parent[i] = subject[i];
+    }
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
 
-    return new Proxy(this, handler);
-  // SlowBuffer.makeFastBuffer(this.parent, this, this.offset, this.length)
+  function createBufferFromBuffer(subject, length = subject.length, offset = 0) {
+    const buffer = Object.create(Buffer.prototype);
+    buffer.length = length;
+    buffer.offset = offset;
+    buffer.parent = subject.parent || new SlowBuffer(length);
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
   }
 
+  Buffer.prototype.values = Buffer.prototype[Symbol.iterator] = function * () {
+    for (let i = 0; i < this.length; i++) {
+      yield this[i];
+    }
+  };
+
+  Buffer.prototype.entries = function * () {
+    for (let i = 0; i < this.length; i++) {
+      yield [i, this[i]];
+    }
+  };
+
+  Buffer.prototype.keys = function * () {
+    for (let i = 0; i < this.length; i++) {
+      yield i;
+    }
+  };
+
   Buffer.prototype._PRINT = function (context) {
     context.output += '<Buffer';
     for (let i = 0; i < Math.min(this.length, 50); i++) {
       context.output += ' ';
-      context.output += this.parent.hexSlice(i, i + 1);
+      context.output += this.parent.hexSlice(this.offset + i, this.offset + i + 1);
```

Comment:
This was a bug in the old PR because of a misunderstanding how our lazy buffers work. Buffers created from other buffers with an offset share the same underlying `SlowBuffer` but use offsets. The representation should respect those offsets to show the correct byte sequences the buffer actually covers.